sympy/integrals/intpoly.py

```diff
@@ -294,21 +321,11 @@ def main_integrate(expr, facets, hp_params, max_degree=None):
                                 (b / norm(a)) / (dim_length + degree)
         return result
     else:
-        polynomials = decompose(expr)
-        for deg in polynomials:
-            poly_contribute = S.Zero
-            facet_count = 0
-            for hp in hp_params:
-                value_over_boundary = integration_reduction(facets,
-                                                            facet_count,
-                                                            hp[0], hp[1],
-                                                            polynomials[deg],
-                                                            dims, deg)
-                poly_contribute += value_over_boundary * (hp[1] / norm(hp[0]))
-                facet_count += 1
-            poly_contribute /= (dim_length + deg)
-            integral_value += poly_contribute
-    return integral_value
+        if not isinstance(expr, list):
+            polynomials = decompose(expr)
+            return _polynomial_integrate(polynomials, facets, hp_params)
+        else:
+            return {exp: _polynomial_integrate(decompose(exp), facets, hp_params) for exp in expr}
```

Comment:
```suggestion
        dec_e = decompose(expr)
        if not isinstance(dec_e, list):
            return _polynomial_integrate(dec_e, facets, hp_params)
        else:
            return {exp: _polynomial_integrate(dec_e, facets, hp_params)
                for exp in expr}
```

Reply:
@smichr I guess your idea is of executing `decompose` only once but here it would treat a list of expressions as a single expression (which was the correction this pr was making in the first case ) .Also if we change `decompose` to check for list and solve for each expression of list internally ,then we cannot solve for polynomials (we can but with added difficulty) .We would need something to differentiate between decompose( [ 1 , x + 3 , x^2 ] ) and decompose( [ 1 , x  , 3 + x^2 ] )  .

Reply:
I think the current code with suggestions by @oscargus is fine (even timewise it runs in almost same time as it takes on master if `max_degree` is given as an input) . I'll resolve this conversation in some time if you don't have any particular doubts .Thanks !